File: festival-web/src/main/java/cn/bdqfork/web/route/response/HtmlResponseHandler.java
package cn.bdqfork.web.route.response;

import cn.bdqfork.web.route.ModelAndView;
import cn.bdqfork.web.route.TemplateManager;
import io.vertx.core.http.HttpServerResponse;
import io.vertx.ext.web.RoutingContext;

/**
 * @author bdq
 * @since 2020/2/21
 */
public class HtmlResponseHandler extends AbstractResponseHandler {
    private TemplateManager templateManager;

    public HtmlResponseHandler(TemplateManager templateManager) {
        this.templateManager = templateManager;
    }

    @Override
    protected void doHandle(RoutingContext routingContext, Object result) throws Exception {
        HttpServerResponse response = routingContext.response();
        ModelAndView modelAndView = (ModelAndView) result;
        String template = templateManager.getTemplatePath() + "/" + modelAndView.getView() + templateManager.getSuffix();
        templateManager.getTemplateEngine().render(modelAndView.getModel(), template, res -> {
            if (res.succeeded()) {
                response.end(res.result());
            } else {
                routingContext.fail(500, res.cause());
            }
        });
    }

}

File: festival-kotlin-web/src/main/kotlin/cn/bdqfork/kotlin/web/route/response/HtmlResponseHandler.kt
package cn.bdqfork.kotlin.web.route.response

import cn.bdqfork.kotlin.web.route.ModelAndView
import cn.bdqfork.kotlin.web.route.TemplateManager
import io.vertx.core.buffer.Buffer
import io.vertx.ext.web.RoutingContext
import io.vertx.kotlin.coroutines.awaitResult
import kotlinx.coroutines.runBlocking

/**
 * @author bdq
 * @since 2020/2/21
 */
class HtmlResponseHandler(private val templateManager: TemplateManager) : AbstractResponseHandler() {
    @Throws(Exception::class)
    override fun doHandle(routingContext: RoutingContext, result: Any) = runBlocking {
        val response = routingContext.response()
        if (templateManager.isEnable) {
            val modelAndView = result as ModelAndView
            val template = templateManager.templatePath + "/" + modelAndView.view + templateManager.suffix
            val buffer = awaitResult<Buffer> { h -> templateManager.templateEngine.render(modelAndView.getModel(), template, h) }
            response.end(buffer)
        } else {
            routingContext.fail(500, IllegalStateException("template is not enabled!"))
        }
    }

}